tests/test_ensemble.py

```diff
@@ -124,3 +126,131 @@ def test_replicated_applications_have_eq_deep_copies_of_parameters(params):
+     pytest.param(_2x2_PARAMS,        _2x2_EXE_ARG,     "step",          1,        1,                 1 , id="Set max permutation to lowest"),
+     pytest.param(_2x2_PARAMS,        _2x2_EXE_ARG,     "step",          6,        2,                 4 , id="Set max permutation, set replicas"),
+     pytest.param(         {},        _2x2_EXE_ARG,     "step",          6,        2,                 4 , id="Set params as dict, set max permutations and replicas"),
+     pytest.param(_2x2_PARAMS,                  {},     "step",          6,        2,                 4 , id="Set params as dict, set max permutations and replicas"),
+     pytest.param(         {},                  {},     "step",          6,        2,                 2 , id="Set params as dict, set max permutations and replicas")
+))
+# fmt: on
+def test_step_strategy(
+    # Parameterized
+    params,
+    exe_arg_params,
+    strategy,
+    max_perms,
+    replicas,
+    expected_num_jobs,
+    # Other fixtures
+    mock_launcher_settings,
+):
+    jobs = Ensemble(
+        "test_ensemble",
+        "echo",
+        ("hello", "world"),
+        file_parameters=params,
+        exe_arg_parameters=exe_arg_params,
+        permutation_strategy=strategy,
+        max_permutations=max_perms,
+        replicas=replicas,
+    ).as_jobs(mock_launcher_settings)
+    assert len(jobs) == expected_num_jobs
+
+
+# fmt: off
+@pytest.mark.parametrize(
+    "                  params,      exe_arg_params,   strategy,  max_perms, replicas, expected_num_jobs",         
+    (pytest.param(_2x2_PARAMS,        _2x2_EXE_ARG,   "random",         30,        1,                16 , id="Set max permutation high"),  # Ask about this 
+     pytest.param(_2x2_PARAMS,                None,   "random",          4,        1,                4  , id="No exe arg params or Replicas"),
+     pytest.param(       None,        _2x2_EXE_ARG,   "random",          4,        1,                4  , id="No Parameters or Replicas"),
+     pytest.param(       None,                None,   "random",          4,        1,                1  , id="No Parameters, Exe_Arg_Param or Replicas"),
+     pytest.param(_2x2_PARAMS,        _2x2_EXE_ARG,   "random",          1,        1,                1  , id="Set max permutation to lowest"),
+     pytest.param(_2x2_PARAMS,        _2x2_EXE_ARG,   "random",          6,        2,                12 , id="Set max permutation, set replicas"),
+     pytest.param(         {},        _2x2_EXE_ARG,   "random",          6,        2,                 8 , id="Set params as dict, set max permutations and replicas"),
+     pytest.param(_2x2_PARAMS,                  {},   "random",          6,        2,                 8 , id="Set params as dict, set max permutations and replicas"),
+     pytest.param(         {},                  {},   "random",          6,        2,                 2 , id="Set params as dict, set max permutations and replicas")
```

Comment:
for all of these sets of parameterized tests, consider adding a `max_params=0` or `max_params=-1` entry. I think it will expose some existing errors in the current impl